Support limit push down for GroupBy

jon-wei opened the pull request:
This patch adds a context flag to the GroupBy query that enables result limiting/sorting at the merge buffer level, allowing the brokers to process less data.

This is accomplished by using a min-max heap with a size limit in the BufferGrouper instead of a list of offsets.

The table buffer in the BufferGrouper is used differently in push down mode: instead of growing as buckets are added, the hash table buffer is split in half. When a half fills, the active table buffer swaps to the other half, and the top N buckets are copied over to the new active buffer.

Note that when the sorting order uses fields that are not in the grouping key, limit push down can result in approximate results.

fjy commented:
@jon-wei any benchmark results?

jon-wei commented:
@fjy 

I ran a benchmark against the TPC-H dataset used by https://github.com/implydata/redshift-benchmark on a small cluster with one r3.2xlarge historical and one r3.2xlarge broker, using the following historical/broker configs:

historical
```
druid.service=druid/historical
druid.port=8083

# HTTP server threads
druid.server.http.numThreads=40

# Processing threads and buffers
#druid.processing.buffer.sizeBytes=536870912
druid.processing.buffer.sizeBytes=1073741824
druid.processing.numThreads=7
druid.processing.numMergeBuffers=4

# Segment storage
druid.segmentCache.locations=[{"path":"/mnt/var/druid/segment-cache","maxSize"\:130000000000}]
druid.server.maxSize=130000000000

# Query cache
druid.historical.cache.useCache=true
druid.historical.cache.populateCache=true
druid.cache.type=local
druid.cache.sizeInBytes=2000000000
```

broker
```
druid.service=druid/broker
druid.port=8082

# HTTP server threads
druid.broker.http.numConnections=5
druid.server.http.numThreads=40

# Processing threads and buffers
#druid.processing.buffer.sizeBytes=536870912
druid.processing.buffer.sizeBytes=1073741824
druid.processing.numThreads=7
druid.processing.numMergeBuffers=4
#druid.broker.http.compressionCodec=GZIP
druid.broker.http.compressionCodec=DEFLATE
#druid.broker.http.compressionCodec=IDENTITY

# Query cache disabled -- push down caching and merging instead
druid.broker.cache.useCache=false
druid.broker.cache.populateCache=false
```

running the following queries (expressed in the RDruid format from the benchmark script):

```
group_by_top_100_parts_v2 <- function(datasource) {
  druid.query.groupBy(
    url = url,
    dataSource   = datasource,
    intervals    = i,
    aggregations = list(
      sum(metric("l_quantity"))
    ),
    granularity = granularity("all"),
    dimensions = list("l_partkey"),
    orderBy=list("l_partkey"),
    limit=100,
    context=list(useCache=F, populateCache=F, groupByStrategy = 'v2', typeHints=list(l_partkey="long")),
    verbose = TRUE,
    benchmark = TRUE
  )
}

group_by_top_100_parts_v2_pushdown_limit <- function(datasource) {
  druid.query.groupBy(
    url = url,
    dataSource   = datasource,
    intervals    = i,
    aggregations = list(
      sum(metric("l_quantity"))
    ),
    granularity = granularity("all"),
    dimensions = list("l_partkey"),
    orderBy=list("l_partkey"),
    limit=100,
    context=list(useCache=F, populateCache=F, groupByStrategy = 'v2', pushDownLimit=T, typeHints=list(l_partkey="long")),
    verbose = TRUE,
    benchmark = TRUE
  )
}
```

with the following results (in seconds):
```
group_by_top_100_parts_v2	261.835591162
group_by_top_100_parts_v2	260.041529968
group_by_top_100_parts_v2_pushdown_limit	202.518083854
group_by_top_100_parts_v2_pushdown_limit	201.167777016
```

Will run more benchmarks next week and try on a larger cluster as well.

himanshug commented on processing/src/main/java/io/druid/query/groupby/GroupByQuery.java:
```diff
@@ -281,10 +291,100 @@ public int compare(Object lhs, Object rhs)
+          AggregatorFactory agg = aggregatorSpecs.get(aggIndex);
+          orderedFieldNames.add(agg.getName());
+          directions.add(direction);
+        }
```
getRowOrderingPushDown(..) is only used when limitSpec does not have aggregator in it, so shoudnt' this case of aggIndex >= 0 should actually be an error

jon-wei replied:
removed the agg checks here since they weren't necessary, thanks

himanshug commented on processing/src/main/java/io/druid/query/groupby/GroupByQuery.java:
```diff
@@ -281,10 +291,100 @@ public int compare(Object lhs, Object rhs)
   }
 
-  public Ordering<Row> getRowOrdering(final boolean granular)
+  public Ordering<Row> getRowOrderingForPushDown(
```
can we mark it private

jon-wei replied:
marked private

himanshug commented on docs/content/querying/groupbyquery.md:
```diff
@@ -214,3 +214,4 @@ When using the "v2" strategy, the following query context parameters apply:
 |`maxMergingDictionarySize`|Can be used to lower the value of `druid.query.groupBy.maxMergingDictionarySize` for this query.|
 |`maxOnDiskStorage`|Can be used to lower the value of `druid.query.groupBy.maxOnDiskStorage` for this query.|
 |`sortByDimsFirst`|Sort the results first by dimension values and then by timestamp.|
+|`pushDownLimit`|When a limit spec is present, enables an optimization that applies the limit at the merge buffer level, reducing the number of results that need to be processed at the broker level. Note that when the sorting order uses fields that are not in the grouping key, this can result in approximate results.|
```
haven't gone through whole code yet... but, speaking as a user, it would be nice for druid to automatically push down limits where results would be exactly same.
and for other case, push limits only if user asked explicitly.

jihoonson replied:
The approximate results seem to mean that the group-by results may not be valid if non-grouping keys are used for sort. This is because of the behavior of ```BufferGrouper```. It first aggregates input rows until its hash table fills. When the hash table fills, it copies the top-n entries found so far to another hash table, and continues aggregation. Here, the copied entires are able to not belong to the real group-by result, and this can make the result invalid.

In this case, we cannot guarantee the approximation ratio because it will depend on the input order. In the worst case, users can get totally unexpected results but it's difficult for them to figure out until running the same query without limit push down. 

So, IMHO, it would be better to disable when non-grouping keys are used for sort. What do you think?

jon-wei replied:
I updated the patch so that the limit push down is enabled when the sorting order only has grouping key fields and noted that within the docs

re: push down when sorting on non-grouping fields, I think it's good to keep it disabled by default, but I also think it's useful to allow it if the user really wants to do it, so I've kept the push down context flag, with a note in the docs that there are no guarantees on the accuracy of the results in that case.

jihoonson commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferGrouper.java:
```diff
@@ -90,14 +120,40 @@
+  // Limit to apply to results.
+  // If limit > 0, track hash table entries in a binary heap with size of limit.
+  // If -1, no limit is applied, hash table entry offsets are tracked with an unordered list with no limit.
+  private int limit;
```
IMO, it would be better to make a new Grouper supporting limit push down because its behavior is quite different from the original one in many parts to achieve a different goal. It will be helpful for better readability and maybe slight performance as well.

himanshug replied:
+1

jon-wei replied:
split the limit handling stuff into a new LimitedBufferGrouper

jon-wei commented on processing/src/main/java/io/druid/query/groupby/GroupByQuery.java:
```diff
@@ -281,10 +291,100 @@ public int compare(Object lhs, Object rhs)
+          }
+      );
+    }
+  };
```
removed ;

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferGrouper.java:
```diff
@@ -241,121 +343,265 @@ public void reset()
+      @Override
+      public Entry<KeyType> next()
+      {
+        return bucketEntryForOffset(wrappedOffsets.get(curr++));
```
Added that check

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferGrouper.java:
```diff
@@ -241,121 +343,265 @@ public void reset()
+      @Override
+      public Entry<KeyType> next()
+      {
+        final int offset = buffer.getInt(tableArenaSize + curr * Ints.BYTES);
```
Added that check

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferGrouper.java:
```diff
@@ -241,121 +343,265 @@ public void reset()
+      public Entry<KeyType> next()
+      {
+        if (curr >= initialHeapSize) {
+          throw new ISE("WTF");
```
changed to throw NSEE

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/ByteBufferMinMaxOffsetHeap.java:
```diff
@@ -0,0 +1,394 @@
+ * (http://portal.acm.org/citation.cfm?id=6621), with some utility functions from
+ * Guava's MinMaxPriorityQueue.
+ */
+public class ByteBufferMinMaxOffsetHeap
```
Hm, I could see that, I'll leave this specialized for the current use case in this PR though, if another use comes up later we can do a PR then

jihoonson replied:
agree.

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/ByteBufferMinMaxOffsetHeap.java:
```diff
@@ -0,0 +1,394 @@
+    }
+  }
+
+  private void printHeap()
```
removed debug print fn

jon-wei commented:
@himanshug @jihoonson 

Thanks for the comments so far, updated the PR

jon-wei commented:
marking this WIP, fixing some test failures, will reopen later

jon-wei commented:
@himanshug @jihoonson 

Updated this with a few more changes:
* added a small optimization to skip aggregation for keys that have already been ejected from the heap when the sorting columns are a subset of the grouping key (can't be in the top N at that point, so no need to aggregate them anymore) 
* changed "has aggregators" condition to "has non-grouping key fields" 
* throws an exception now if the user sets the flag to force push down when post aggs are included in the sorting order (needs a change to the post agg interface to support byte serde and something like getMaxIntermediateSize() from AggregatorFactory, and I think that's too big for this PR)

himanshug commented on processing/src/main/java/io/druid/query/groupby/GroupByQuery.java:
```diff
@@ -281,10 +290,93 @@ public int compare(Object lhs, Object rhs)
+  private Ordering<Row> getRowOrderingForPushDown(
+      final boolean granular,
+      final DefaultLimitSpec limitSpec,
+      final List<AggregatorFactory> aggregatorSpecs
```
unused

himanshug commented on processing/src/main/java/io/druid/query/groupby/GroupByQuery.java:
```diff
@@ -370,6 +462,21 @@ private static int compareDims(List<DimensionSpec> dimensions, Row lhs, Row rhs)
     return 0;
   }
 
+  private static int compareDimsAndAggs(List<String> fields, List<Integer> directions, Row lhs, Row rhs)
```
`compareDims` instead?

himanshug commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/SpillingGrouper.java:
```diff
@@ -75,24 +77,42 @@ public SpillingGrouper(
-    );
+    this.keyObjComparator = keySerdeFactory.objectComparator(false);
+    this.defaultOrderKeyObjComparator = keySerdeFactory.objectComparator(true);
+    if (limit > -1) {
```
is limit == 0 a valid value ?

himanshug replied:
if it is, let us change things so that limit should be a positive integer... limit <= 0 makes no sense

himanshug commented on processing/src/main/java/io/druid/query/groupby/GroupByQuery.java:
```diff
@@ -281,10 +290,93 @@ public int compare(Object lhs, Object rhs)
+    final boolean pushDownSort = getContextBoolean(GroupByQueryConfig.CTX_KEY_PUSH_DOWN_LIMIT, false);
+    if (pushDownSort && !DefaultLimitSpec.sortingOrderHasNonGroupingFields(limitSpec, dimensions)) {
+      return getRowOrderingForPushDown(granular, (DefaultLimitSpec) limitSpec, aggregatorSpecs);
+    }
```
so , here it is expected limits are pushed only when pushDownSort is explicitly set by the user in the request.. while in RowBasedGrouperHelper it will happen automatically if limit was non-negative and ordering columns were subset of dimensions, wouldn't this produce wrong results if user did not enable pushDown but it was automatic.... am i missing something?
also, "whether limit is pushed down or not" code can be refactored into a common place and is used here and RowBasedGrouperHelper so that no disparity can happen.

himanshug commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/LimitedBufferGrouper.java:
```diff
@@ -0,0 +1,627 @@
+import java.util.List;
+import java.util.NoSuchElementException;
+
+public class LimitedBufferGrouper<KeyType> implements Grouper<KeyType>
```
I was imagining LimitedBufferGrouper and BufferGrouper to share common code so that if we ever identify any bugs then we need to fix in one place, is that feasible and makes sense?

himanshug commented:
i looked at LimitedBufferGrouper and ByteBufferMinMaxOffsetHeap , they look OK but I'm assuming enough unit testing is done to catch any possible issues.

himanshug commented on docs/content/querying/groupbyquery.md:
```diff
@@ -214,3 +214,4 @@ When using the "v2" strategy, the following query context parameters apply:
 |`maxMergingDictionarySize`|Can be used to lower the value of `druid.query.groupBy.maxMergingDictionarySize` for this query.|
 |`maxOnDiskStorage`|Can be used to lower the value of `druid.query.groupBy.maxOnDiskStorage` for this query.|
 |`sortByDimsFirst`|Sort the results first by dimension values and then by timestamp.|
+|`pushDownLimit`|Limits/orderbys are applied at the merge buffer level (only the top N results are kept in the merge buffers, reducing the workload at the broker level), when all fields in the orderby are part of the grouping key. When the sorting order uses fields that are not in the grouping key, this can result in approximate results with unknown accuracy, so this optimization is disabled by d ...
```
i don't think user needs to know about the "merge buffer stuff"... it would suffice to say that...

broker node always pushes down limits to historical nodes, when it is guaranteed to not alter the overall results i.e.  when all fields in the orderby...



on a side note, similar to topN, we can improve accuracy of the aggregator case if we inflate the limit to say 100x than what user specified when pushing to historicals.

jon-wei replied:
@himanshug changed docs to not mention the merge buffers

@nishantmonu51 If the fields in the order by are all part of the grouping key but pushDownLimit is explicitly set to false, the the limit push down will still happen.

Renamed this to forcePushDownLimit

jon-wei commented:
pushing this to 10.1

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/LimitedBufferGrouper.java:
```diff
@@ -0,0 +1,627 @@
+        if (curr >= initialHeapSize) {
+          throw new NoSuchElementException();
+        }
+        final int offset = offsetHeap.removeMin();
```
Hm, not sure if there's a way to get a sorted iterator from the heap non-destructively, an alternative would be build a list and sort it, but that introduces sorting overhead as well, they should both be N * logN operations on average

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java:
```diff
@@ -373,21 +532,51 @@ private static int compareDimsInRows(RowBasedKey key1, RowBasedKey key2, int dim
+          );
+        }
+        if (cmp != 0) {
+          return cmp * directions.get(i);
```
Changed these to use booleans/swapped directions

jon-wei commented on processing/src/main/java/io/druid/query/groupby/GroupByQuery.java:
```diff
@@ -370,6 +462,21 @@ private static int compareDims(List<DimensionSpec> dimensions, Row lhs, Row rhs)
+          rhs.getRaw(fieldName)
+      );
+      if (dimCompare != 0) {
+        return dimCompare * directions.get(i);
```
When limit push down is not used, it will just sort by the default order (all dims, no aggs)

jon-wei commented:
@himanshug @jihoonson @nishantmonu51 

Updated this with some large refactoring changes:
- Made a base AbstractBufferGrouper with different implementations for non-pushdown and push down
- Split out the off-heap hash table, list, and min-max heap structures into separate classes

Waiting on benchmark results, will update this PR with them when ready.

jihoonson commented:
Great! Thanks for your effort. I'll review soon.

jon-wei commented:
Added benchmark results to my second comment in this PR's thread

jon-wei commented on processing/src/test/java/io/druid/query/groupby/GroupByQueryRunnerTest.java:
```diff
@@ -8145,4 +8145,709 @@ public void testGroupByWithAggsOnNumericDimensions()
+    );
+    Map<String, Object> context = Maps.newHashMap();
+    List<Row> allGranExpectedResults = Arrays.asList(
+        GroupByQueryRunnerTestHelper.createExpectedRow("2011-04-02", "alias", "travel", "rows", 2L, "idx", 243L),
```
I believe the test is correct, the query has ALL granularity with an interval that covers 4-02 to 4-04, so the entries from 4-03 are also included, which is why the aggregation results are doubled.

I copied this test to master and ran it there and got the same results to confirm.

jihoonson replied:
Oh, sorry. You are right. My mistake.

jon-wei commented on processing/src/main/java/io/druid/query/groupby/GroupByQuery.java:
```diff
@@ -300,10 +317,172 @@ public int compare(Object lhs, Object rhs)
+      DefaultLimitSpec defaultLimitSpec = (DefaultLimitSpec) limitSpec;
+
+      // If only applying an orderby without a limit, don't try to push down
+      if (defaultLimitSpec.getLimit() == Integer.MAX_VALUE) {
```
`validateAndGetForceLimitPushDown()` only throws the exception if the user explicitly sets `forceLimitPushDown` to true in the query context.

In this block, `forceLimitPushDown` may be false, so it's checking for the case where a limit was not provided (e.g., the user only wants to change the sort, but not impose a limit, without specifying forced push down)

jon-wei replied:
hm, somewhat unrelated, maybe it's better to represent "no limit specified" as null instead of Integer.MAX_VALUE, I'll think about changing that

jon-wei commented on processing/src/main/java/io/druid/query/groupby/GroupByQuery.java:
```diff
@@ -300,10 +317,172 @@ public int compare(Object lhs, Object rhs)
+    return false;
+  }
+
+  private Ordering<Row> getRowOrderingForPushDown(
```
When limit push down is applied, the partial results would be sorted by the ordering specified by the limit/order spec (unlike non-push down case where the results always use the default natural ascending order), so when merging these partial result streams, the merge needs to use the same ordering to get correct results

jon-wei replied:
You can see the effect of disabling the custom ordering with `testMergeResultsWithLimitPushDown`

jihoonson replied:
Ah I got it. Thanks.

himanshug replied:
@jon-wei please add this as comment.

jon-wei replied:
@himanshug Added this as a comment

jon-wei commented on processing/src/main/java/io/druid/query/BaseQuery.java:
```diff
@@ -206,7 +206,7 @@ public boolean getContextBoolean(String key, boolean defaultValue)
   }
 
-  protected Map<String, Object> computeOverridenContext(Map<String, Object> overrides)
+  public Map<String, Object> computeOverridenContext(Map<String, Object> overrides)
```
Changed this back to protected, adjusted call site to use `withOverriddenContext()`

jon-wei commented on processing/src/main/java/io/druid/query/groupby/GroupByQuery.java:
```diff
@@ -300,10 +317,172 @@ public int compare(Object lhs, Object rhs)
+    final Set<Integer> dimsInOrderBy = new HashSet<>();
+    final List<Boolean> needsReverseList = new ArrayList<>();
+    final List<Boolean> isNumericField = Lists.newArrayList();
+    final List<StringComparator> comparators = Lists.newArrayList();
```
Made the list creation pattern consistent

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/RowBasedGrouperHelper.java:
```diff
@@ -91,20 +98,35 @@
         )
     );
+
+    final boolean willApplyLimitPushDown = query.willApplyLimitPushDown();
```
Changed this to read the final field instead of recomputing

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/ByteBufferIntList.java:
```diff
@@ -0,0 +1,75 @@
+public class ByteBufferIntList
+{
+  private ByteBuffer buffer;
+  private int maxElements;
```
Added a maxElements check to add, made this final along with the buffer

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/GroupByRowProcessor.java:
```diff
@@ -160,6 +160,7 @@ public ByteBuffer get()
                   aggregatorFactories
               );
               final Grouper<RowBasedKey> grouper = pair.lhs;
+             // grouper.init();
```
removed

jon-wei commented on processing/src/main/java/io/druid/query/groupby/strategy/GroupByStrategyV2.java:
```diff
@@ -228,65 +226,71 @@ public boolean doMergeResults(final GroupByQuery query)
-              public Row apply(final Row row)
-              {
-                // Apply postAggregators and fudgeTimestamp if present and if this is the outermost mergeResults.
+    Map<String, Object> newContext = query.computeOverridenContext(
```
Changed this to use `withOverriddenContext()`

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/Grouper.java:
```diff
@@ -231,6 +235,18 @@ public String toString()
+     *
+     * @return comparator for keys + aggs
+     */
+    KeyComparator bufferComparatorWithAggregators(AggregatorFactory[] aggregatorFactories, int[] aggregatorOffsets);
```
I renamed this to `BufferComparator`.

I think making a `BufferComparatorFactory` would be problematic for the `RowBasedKeySerde`, the comparisons on dimensions rely on the dictionary built by the KeySerde, there would be a lot of redundant information between the BufferComparatorFactory and the KeySerde (e.g., the serdeHelpers, timestamp handling information, dictionaries). 

I suppose I view KeySerde as a sort of BufferComparatorFactory, and I feel like the additional withAggregators comparator method isn't inappropriate. 

It's generating a comparator that deals with serialized key portions, interleaved with some extra aggregator comparisons, so the KeySerde is still dealing with its core function of handling operations related to serialized keys.

jon-wei commented:
The latest commit also adds a new `BucketUpdateHandler` interface used by the hash table, to handle updates to the bucket offset list and the aggregator relocation introduced in https://github.com/druid-io/druid/pull/4071

jon-wei commented:
@jihoonson Addressed your comments, thanks for the review!

@himanshug @nishantmonu51 Can you please review this PR again when you have a chance?

jon-wei commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/ByteBufferIntList.java:
```diff
@@ -0,0 +1,75 @@
+
+import java.nio.ByteBuffer;
+
+public class ByteBufferIntList
```
Hm, I'm leaning towards keeping this as-is for now, since it's only being used for a narrow use case by the BufferGrouper right now and the AbstractList has quite a few methods to implement, I feel like it would be too large for this PR, without being necessary for the push down changes

jihoonson commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/LimitedBufferGrouper.java:
```diff
@@ -0,0 +1,538 @@
+
+  private class AlternatingByteBufferHashTable extends ByteBufferHashTable
+  {
+    private ByteBuffer totalHashTableBuffer;
```
Unused variable.

himanshug commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/AbstractBufferGrouper.java:
```diff
@@ -0,0 +1,196 @@
+
+  protected ByteBufferHashTable hashTable;
+  protected ByteBuffer hashTableBuffer; // buffer for the entire hash table (total space, not individual growth)
+
```
can you please add comment on why above variables are not final? (I think because you can reset() and reuse the same Grouper )

gianm replied:
Some look like they should be final, maybe not all of them though.

jon-wei replied:
Added comments on why the non-final variables are not final

himanshug commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/AbstractBufferGrouper.java:
```diff
@@ -0,0 +1,196 @@
+      for (int i = 0; i < aggregators.length; i++) {
+        aggregators[i].init(tableBuffer, bucketStartOffset + aggregatorOffsets[i]);
+      }
+
```
nit: interesting that we encapsulate placing the key via a method on hashTable but aggregates are updated directly. and not using same approach for both.

jon-wei replied:
I went with this approach since assumptions about the key+hash layout in the bucket are built into how the hash table operates, but the aggregators are in the "value" portion of a bucket which I view as being opaque to the hash table implementation

himanshug commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/LimitedBufferGrouper.java:
```diff
@@ -0,0 +1,536 @@
+    }
+    this.totalBuffer = bufferSupplier.get();
+
+    LimitedBufferGrouper.validateBufferCapacity(
```
can validateBufferCapacity be private and called directly instead of qualifying with classname? It just looks like a private utility method to me.

jon-wei replied:
Changed to private utility method

himanshug commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/ByteBufferHashTable.java:
```diff
@@ -0,0 +1,380 @@
+  protected int maxSize;
+
+  // current number of available/used buckets in the table
+  protected int buckets;
```
can we call this `maxPossibleBuckets` ?

himanshug replied:
may be also document in the comment for above two variables that these numbers change on table resize.

himanshug replied:
`maxBuckets` is probably better.

jon-wei replied:
Renamed to `maxBuckets` and added comments re: table resize

himanshug commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/ByteBufferHashTable.java:
```diff
@@ -0,0 +1,380 @@
+  protected int size;
+
+  // Maximum number of elements in the table before it must be resized
+  protected int maxSize;
```
can we call this `regrowthThreshold` ?

jon-wei replied:
renamed to `regrowthThreshold`

himanshug commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/ByteBufferHashTable.java:
```diff
@@ -0,0 +1,380 @@
+  protected ByteBuffer buffer;
+  protected int bucketSizeWithHash;
+  protected int tableArenaSize;
+  protected int keySize;
```
last 4 variables can be final i think

jon-wei replied:
made these final

himanshug commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/ByteBufferHashTable.java:
```diff
@@ -0,0 +1,380 @@
+
+import java.nio.ByteBuffer;
+
+public class ByteBufferHashTable
```
@gianm is concept of regrowth there only because at initializing time we need to set 0 in first byte of all the buckets and it would be too expensive to do that upfront for buffers ? if yes, then, it'll be interesting to see how Unsafe.setMemory(..) performs over 1 GB buffer and if that is fast enough then regrowth business could possibly be removed.

gianm replied:
Yes, that's the reason. When I benchmarked "simple" groupBys with small result sets, zeroing out the first byte of each bucket was a big performance hog.

jon-wei replied:
That seems worth looking into, though I think that would take some time/effort to evaluate and could be handled in a follow-on PR.

himanshug commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/LimitedBufferGrouper.java:
```diff
@@ -0,0 +1,536 @@
+  {
+    private ByteBuffer[] subHashTableBuffers;
+    private ByteBufferHashTable[] subHashTables;
+    private ByteBufferHashTable activeHashTable;
```
not sure why we are creating ByteBufferHashTable structures on two haves? can't we just keep two ByteBuffer for two halves and a reference to activeByteBuffer ?

also, may be some comments saying there are only two of these (them being array gave me impression that there were many of those) .... alternated on each regrow where elements beyond limit are discarded.

jon-wei replied:
Changed this to use two ByteBuffers directly without the sub HashTables, added a comment on there being two buffers and a description of the swapping

himanshug commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/LimitedBufferGrouper.java:
```diff
@@ -0,0 +1,536 @@
+  // Limit to apply to results.
+  // If limit > 0, track hash table entries in a binary heap with size of limit.
+  // If -1, no limit is applied, hash table entry offsets are tracked with an unordered list with no limit.
+  private int limit;
```
what is the case where `limit <= 0` is valid?

jon-wei replied:
This was an outdated comment from an older revision, limit will never be <= 0 now, I deleted the comment

himanshug commented on processing/src/main/java/io/druid/query/groupby/GroupByQuery.java:
```diff
@@ -300,10 +323,172 @@ public int compare(Object lhs, Object rhs)
+    return forcePushDown;
+  }
+
+  public boolean determineApplyLimitPushDown()
```
i think this can be implemented in one line `return validateAndGetForceLimitPushDown() || DefaultLimitSpec.sortingOrderHasNonGroupingFields(..);`

you might have to add the  `defaultLimitSpec.getLimit() == Integer.MAX_VALUE` check inside `DefaultLimitSpec.sortingOrderHasNonGroupingFields(..)`

jon-wei replied:
Hm, I decided to change this area in a different way, I removed an unnecessary null and type check from sortingOrderHasNonGroupingFields(), but kept determineApplyLimitPushDown() the same.

The sortingOrder method is used elsewhere, where the MAX_VALUE check is no longer relevant (as it would've been already checked earlier when the query object was created), so I didn't think it really belongs there.

 I also felt that the conditions for when limit push down will be applied are more clear if they're all in the determineApplyLimitPushDown() method.

himanshug commented:
@jon-wei  looks good overall besides some comments. it makes sense overall , but hard to identify bugs. hoping that you have tested it on some cluster with few different nested/non-nested groupBy queries.

jon-wei commented:
@himanshug 

I've addressed your comments, thanks for the review!

Re: testing, the patch has been tested with non-nested groupBys on a large TPC-H dataset as part of the benchmarking. I am currently in the process of doing more stress testing on this with more query variations like nested GroupBys, I'll make a comment here once that testing is done.

himanshug commented on processing/src/main/java/io/druid/query/groupby/epinephelinae/LimitedBufferGrouper.java:
```diff
@@ -0,0 +1,520 @@
+      // clear the used bits of both buffers
+      for (int i = 0; i < maxBuckets; i++) {
+        subHashTableBuffers[0].put(i * bucketSizeWithHash, (byte) 0);
+        subHashTableBuffers[1].put(i * bucketSizeWithHash, (byte) 0);
```
this one is not needed and will automatically be zeroed out in adjustTableWhenFull(..) when needed.

jon-wei replied:
Got rid of the second buffer reset there

jon-wei commented:
@himanshug 

I tested the patch with some nested group bys against the TPC-H dataset that I benchmarked against earlier, the results are correct.

I've deployed the patch to Imply's internal test cluster as well.

Also merged master just now

himanshug commented:
thanks @jon-wei :+1:

fjy commented:
👍

leventov commented on processing/src/main/java/io/druid/query/groupby/GroupByQueryConfig.java:
```diff
@@ -66,6 +67,12 @@
+  private boolean forcePushDownLimit = false;
+
+  @JsonProperty
+  private Class<? extends GroupByQueryMetricsFactory> queryMetricsFactory;
```
@jon-wei could you please remove this field, getter and setter? They were removed here: https://github.com/druid-io/druid/pull/4336/files#diff-5e30ea112240e82f233099071bb3389e but resurrected in this PR.

jon-wei replied:
@leventov Patch to remove those is here, https://github.com/druid-io/druid/pull/4383, thanks for catching that

leventov commented on processing/src/main/java/io/druid/query/groupby/GroupByQuery.java:
```diff
@@ -97,6 +102,8 @@ public static Builder builder()
   private final List<AggregatorFactory> aggregatorSpecs;
   private final List<PostAggregator> postAggregatorSpecs;
 
+  private final Function<Sequence<Row>, Sequence<Row>> limitFn;
```
This field is unused.
